feat: Add flutter integration driver commands and tests

MummanaSubramanya opened the pull request:
Addressing https://github.com/AppiumTestDistribution/appium-flutter-integration-driver/issues/52

KazuCocoa commented:
Could you add tests in unit test as possible instead of functional test?
https://github.com/appium/python-client/tree/master/test/unit


As a client, what http request should be sent is important but it does not need to test driver side.



Existing functional tests include tests for appium project's drivers themselves. They would be removed as not necessary for client tests.

KazuCocoa commented:
Could you use `flutter_intergation` namespace or `flutter/flutter_integration`?

KazuCocoa commented on appium/webdriver/extensions/flutter/flutter_commands.py:
```diff
@@ -0,0 +1,166 @@
+        Returns:
+            str: Image ID of the injected image.
+        """
+        import os
```
I think this import can be on the top of this file instead of in method.

MummanaSubramanya replied:
Thanks! Fixed

KazuCocoa commented on appium/webdriver/flutter_finder.py:
```diff
@@ -0,0 +1,51 @@
+        return FlutterFinder(FlutterBy.FLUTTER_TEXT_CONTAINING,  value)
+    
+    def to_dict(self):
+        return { "using": self.using, "value": self.value} 
```
Maybe this double-quote will get lint error when we run CI. This should be single quote.

MummanaSubramanya replied:
Thanks! Fixed

KazuCocoa commented on appium/webdriver/extensions/flutter/flutter_commands.py:
```diff
@@ -0,0 +1,166 @@
+        """
+        self.execute_flutter_command("activateInjectedImage", {'imageId': image_id})
+        
+    def execute_flutter_command(self, scriptName: str, params: dict) -> Any:
```
Is this a file private method?

```suggestion
    def _execute_flutter_command(self, scriptName: str, params: dict) -> Any:
```

MummanaSubramanya replied:
No I dnt want it as a private method. I have added the documentation.

KazuCocoa commented:
I left some comments.

- convert tests to unit test to focus on each client method's input/output (fast and less flakiness as client)
    - Existing unit test should help to see how to check the http request's input/output as a client
- name space
- some code improvement

Thank you for the contribution, btw :)

mykola-mokhnach commented on appium/options/flutter/flutter_element_wait_timeout_option.py:
```diff
@@ -0,0 +1,33 @@
+        return self.get_capability(FLUTTER_ELEMENT_WAIT_TIMEOUT)
+
+    @flutter_element_wait_timeout.setter
+    def flutter_element_wait_timeout(self, time_in_millis: int) -> None:
```
Please change the variable type to Union[int, timedelta], like it is done for other similar time-related capabilities. See, for example, https://github.com/appium/python-client/blob/master/appium/options/android/uiautomator2/uiautomator2_server_install_timeout_option.py

mykola-mokhnach commented on appium/options/flutter/flutter_element_wait_timeout_option.py:
```diff
@@ -0,0 +1,33 @@
+class FlutterElementWaitTimeOutOption(SupportsCapabilities):
+    
+    @property
+    def flutter_element_wait_timeout(self) -> Optional[int]:
```
Please add docstrings like in other capability classes

mykola-mokhnach replied:
same comment to other option classes

MummanaSubramanya replied:
Thanks! Updated

mykola-mokhnach commented on appium/webdriver/extensions/flutter/flutter_commands.py:
```diff
@@ -0,0 +1,166 @@
+        
+     # wait commands   
+     
+    def wait_for_visible(self, locator: Union[WebElement, FlutterFinder], time_out: Optional[int] = None) -> None:
```
in python seconds are represented by float values, not integers

MummanaSubramanya replied:
Thanks! Updated

mykola-mokhnach commented on appium/webdriver/extensions/flutter/flutter_commands.py:
```diff
@@ -0,0 +1,166 @@
+            None: 
+        """
+        if isinstance(locator, WebElement):
+            opts = {'element': locator, 'timeout': time_out}
```
should we actually pass the element object itself or its identifier?

MummanaSubramanya replied:
it handled within appium-flutter-server https://github.com/AppiumTestDistribution/appium-flutter-server/blob/0019f6ebf242c78c211872d8c64744f3582bc8b7/server/lib/src/handler/wait/base_wait.dart#L21

mykola-mokhnach replied:
I an not sure it is going to work like that. Could you please point me to a passing e2e test that verifies this approach?

MummanaSubramanya replied:
here is the code https://github.com/appium/python-client/pull/1022/files#diff-025756cb750b47a876194b14a36ff2ebf8236b98a581eb1c7c1249873425a87bR40 and the same has been implemented in java client too https://github.com/appium/java-client/blob/9b3194e29857b6204f51d7f198d3583d2d904e50/src/main/java/io/appium/java_client/flutter/commands/WaitParameter.java#L30

mykola-mokhnach replied:
thanks. I mean I would like to check the github actions log where the test has passed

mykola-mokhnach replied:
Currently all flutter tests fail with https://github.com/appium/python-client/actions/runs/10966892515/job/30463953515?pr=1022#step:17:76

You may verify they work as expected by running tests in the local fork

mykola-mokhnach commented on appium/webdriver/extensions/flutter/flutter_commands.py:
```diff
@@ -0,0 +1,166 @@
+            None:
+        """
+        opts = {'origin': element}
+        if offset:
```
prefer to use explicit `is not None` comparison

MummanaSubramanya replied:
Thanks! Updated

mykola-mokhnach commented on appium/webdriver/extensions/flutter/flutter_commands.py:
```diff
@@ -0,0 +1,166 @@
+        """
+        self.execute_flutter_command('dragAndDrop', {'source': source, 'target': target})
+        
+    def scroll_till_visible(self, scroll_to: FlutterFinder, scroll_direction: Optional[str] = 'down', **opts: Any) -> WebElement:
```
Prefer to use explicit enum there. See https://docs.python.org/3/library/enum.html

MummanaSubramanya replied:
Thanks! Added

mykola-mokhnach commented on appium/webdriver/extensions/flutter/flutter_commands.py:
```diff
@@ -0,0 +1,166 @@
+        Returns:
+            None:
+        """
+        self.execute_flutter_command("activateInjectedImage", {'imageId': image_id})
```
Please use single quotes everywhere

MummanaSubramanya replied:
Thanks! Updated

mykola-mokhnach commented on appium/webdriver/flutter_finder.py:
```diff
@@ -0,0 +1,51 @@
+    def by_flutter_text_containing(value: str) -> 'FlutterFinder':
+        return FlutterFinder(FlutterBy.FLUTTER_TEXT_CONTAINING,  value)
+    
+    def to_dict(self):
```
Please add the return type

MummanaSubramanya replied:
Thanks! Updated

mykola-mokhnach commented on test/functional/flutter/commands_test.py:
```diff
@@ -0,0 +1,125 @@
+        message_field_finder = FlutterFinder.by_flutter_key('message_field')
+        toggle_button_finder = FlutterFinder.by_flutter_key('toggle_button')
+        
+        message_field = self.driver.find_element(*message_field_finder.as_strings())
```
as_strings -> to_args

MummanaSubramanya replied:
Thanks! Updated

mykola-mokhnach commented on test/functional/flutter/commands_test.py:
```diff
@@ -0,0 +1,125 @@
+
+        toggle_button.click()
+        self.flutter_command.wait_for_invisible(message_field_finder)
+        assert len(self.driver.find_elements(*message_field_finder.as_strings())) == 0
```
it probably makes sense to create a wrapper over find_elements, so we may directly pass Finder instance to it to avoid unnecessary duplication

MummanaSubramanya replied:
For now I think its not required. I have just wrote it as an example  **if** we want to use finder instance.

mykola-mokhnach commented on appium/webdriver/extensions/flutter/flutter_commands.py:
```diff
@@ -0,0 +1,166 @@
+from appium.webdriver.webelement import WebElement
+
+
+class FlutterCommand():
```
This could be better done as a mixin, so one may use it as

class MyDriver(IOSDriver, FlutterCommandsMixin):
....

MummanaSubramanya replied:
> This could be better done as a mixin, so one may use it as
> 
> class MyDriver(IOSDriver, FlutterCommandsMixin): ....

Are you suggesting to rename the class as `FlutterCommandMixin` ?

mykola-mokhnach replied:
nvm, lets keep it as is

mykola-mokhnach commented on test/functional/flutter/helper/test_helper.py:
```diff
@@ -0,0 +1,55 @@
+        self.driver = webdriver.Remote(SERVER_URL_BASE, options=flutterOptions.load_capabilities(desired_caps))
+        self.flutter_command = FlutterCommand(self.driver)
+        if is_ci():
+            self.driver.start_recording_screen()
```
where do we use these screen recordings?

MummanaSubramanya replied:
Thanks I copied the exisiting android functional tests code. I think we dnt need that

MummanaSubramanya commented:
> Could you use `flutter_intergation` namespace or `flutter/flutter_integration`?

updated

MummanaSubramanya commented on appium/webdriver/common/flutterby.py:
```diff
@@ -0,0 +1,25 @@
+
+from selenium.webdriver.common.by import By
+
+class FlutterBy(By):
```
Thanks.
Agreed! I think Its better to have a prefix `FLUTTER_INTEGRATION` as we also have `appium flutter driver.` Now we can use it like `AppiumBy.FLUTTER_INTEGRATION_KEY`,`AppiumBY.FLUTTER_INTEGRATION_TEXT`

MummanaSubramanya commented on appium/webdriver/flutter_finder.py:
```diff
@@ -0,0 +1,51 @@
+
+from appium.webdriver.common.flutterby import FlutterBy
+
+class FlutterFinder():
```
Thanks! Fixed.

mykola-mokhnach commented on appium/options/flutter_integration/base.py:
```diff
@@ -0,0 +1,40 @@
+# specific language governing permissions and limitations
+# under the License.
+
+from typing import Dict
```
please apply black and isort formatting

MummanaSubramanya replied:
thanks! Applied for all files

mykola-mokhnach commented on appium/webdriver/extensions/flutter_integration/scroll_directions.py:
```diff
@@ -0,0 +1,8 @@
+    UP = 'up'
+    DOWN = 'down'
+    
+    def as_string(self):
```
this method is not really needed. You may use `.value`

MummanaSubramanya replied:
Thanks! fixed

mykola-mokhnach replied:
was the change committed?

MummanaSubramanya replied:
Yes .. I forgot to push the changes

mykola-mokhnach commented on appium/webdriver/flutter_finder.py:
```diff
@@ -0,0 +1,51 @@
+    def to_dict(self) -> dict:
+        return { 'using': self.using, 'value': self.value} 
+    
+    def as_args(self) -> str:
```
tuple[str, str]

MummanaSubramanya replied:
Thanks! fixed

mykola-mokhnach commented on test/functional/flutter_integration/commands_test.py:
```diff
@@ -0,0 +1,126 @@
+    def test_wait_command(self):
+        self.__open_screen('Lazy Loading')
+    
+        message_field_finder = FlutterFinder.by_flutter_key('message_field')
```
not sure if we need to repeat `flutter` twice. Maybe rename `FlutterFinder.by_flutter_key` to `FlutterFinder.by_key`?

MummanaSubramanya replied:
Agreed! Updated.

mykola-mokhnach commented on test/functional/flutter_integration/commands_test.py:
```diff
@@ -0,0 +1,126 @@
+        assert len(self.driver.find_elements(*message_field_finder.as_args())) == 0
+
+        toggle_button.click()
+        self.flutter_command.wait_for_visible(message_field)
```
if we had it implemented as a mixin then we could call flutter commands as 

```
driver.flutter_wait_for_visible(...)
assert len(self.driver.find_flutter_elements(message_field_finder)) == 1
```

mykola-mokhnach commented on test/functional/flutter_integration/commands_test.py:
```diff
@@ -0,0 +1,126 @@
+        assert len(self.driver.find_elements(*message_field_finder.as_args())) == 1
+     
+    def test_scroll_till_visible_command(self):
+        self.__open_screen('Vertical Swiping')
```
use single underscore to mark protected methods

MummanaSubramanya replied:
its a private method only used in this class

mykola-mokhnach commented on appium/webdriver/extensions/flutter_integration/flutter_commands.py:
```diff
@@ -0,0 +1,179 @@
+from appium.webdriver.webelement import WebElement
+
+
+class FlutterCommand:
```
How about renaming this class to FlutterExtensions?

MummanaSubramanya replied:
FlutterExtensions sounds more generic and doesn't immediately convey that it's about commands or actions.
I would rather like to stick with same name or `FlutterActions`. Please suggest

mykola-mokhnach commented on appium/webdriver/extensions/flutter_integration/flutter_commands.py:
```diff
@@ -0,0 +1,179 @@
+                scrollView (str): The view of the scroll.
+                delta (int): delta for the scroll
+                maxScrolls (int): Max times to scroll
+                settleBetweenScrollsTimeout (int): settle timeout
```
Please mention explicitly the time unit used there (I assume milliseconds) and default values if not provided

MummanaSubramanya replied:
fixed and added the default values

mykola-mokhnach replied:
hm, I still cannot find neither for KeywordArgs

MummanaSubramanya replied:
I have pushed the changes.

mykola-mokhnach commented on appium/webdriver/extensions/flutter_integration/flutter_commands.py:
```diff
@@ -0,0 +1,179 @@
+        Scrolls until the specified element becomes visible.
+
+        Args:
+            scroll_to (FlutterFinder): The Flutter element to scroll to.
```
there is no need to repeat types in doctrings

MummanaSubramanya replied:
are you suggesting me to remove `(FlutterFinder)`?

mykola-mokhnach replied:
yes, do it for all args where the time is set explicitly upon declaration

MummanaSubramanya replied:
Any reason ? I see it is useful when I see documentation 
<img width="518" alt="Screenshot 2024-09-21 at 6 55 05 PM" src="https://github.com/user-attachments/assets/5bfd4ed7-ad61-4e6c-8e4b-e55e39c68038">

KazuCocoa commented on appium/webdriver/common/appiumby.py:
```diff
@@ -25,3 +25,8 @@ class AppiumBy(By):
     ACCESSIBILITY_ID = 'accessibility id'
     IMAGE = '-image'
     CUSTOM = '-custom'
+    FLUTTER_INTEGRATION_SEMANTICS_LABEL = '-flutter semantics label'
```
👍 

Can you leave a link as a comment for those as not in appium org repo for future maintenance?

MummanaSubramanya replied:
Thanks. Could you please clarify link I should add in comment .

KazuCocoa replied:
I think the driver project's GitHub repo is good

MummanaSubramanya replied:
done

mykola-mokhnach commented on appium/webdriver/extensions/flutter_integration/flutter_commands.py:
```diff
@@ -0,0 +1,185 @@
+        else:
+            opts['locator'] = locator.to_dict()
+        if time_out is not None:
+            opts['timeout'] = time_out
```
does the downstream API also accept float seconds or should we provide integer milliseconds instead?

MummanaSubramanya replied:
here is the code https://github.com/AppiumTestDistribution/appium-flutter-server/blob/0019f6ebf242c78c211872d8c64744f3582bc8b7/server/lib/src/models/api/wait.dart#L28 it accpets value as a nullable num (which can be either an int or a float)

mykola-mokhnach commented on appium/webdriver/extensions/flutter_integration/flutter_commands.py:
```diff
@@ -0,0 +1,185 @@
+            None:
+        """
+        opts: Dict[str, Union[WebElement, Dict[str, str], float]] = {}
+        if isinstance(locator, WebElement):
```
this condition is duplicated multiple times. Please extract it to a function

MummanaSubramanya replied:
Thanks updated

KazuCocoa commented on appium/webdriver/flutter_finder.py:
```diff
@@ -0,0 +1,53 @@
+from appium.webdriver.common.appiumby import AppiumBy
+
+
+class FlutterFinder:
```
I think this class would be nice to be in `appium/webdriver/extensions/flutter_integration`... since this is only for the flutter integration driver specific

MummanaSubramanya replied:
Thanks! updated

KazuCocoa commented on .github/workflows/functional-test.yml:
```diff
@@ -196,3 +196,145 @@ jobs:
         name: appium-android-${{matrix.test_targets.name}}.log
         path: appium.log
+
+  flutter_e2e_test:
```
If the unit test was sufficient as this client-side implementation's test, this `flutter_e2e_test` can be removed from this PR as redundant slower tests

mykola-mokhnach replied:
Flutter e2e tests are executed in parallel with others. They should not slow down the main flow.

MummanaSubramanya replied:
@KazuCocoa As mentioned by @mykola-mokhnach  All e2e tests executed in parallel. Also having e2e tests would be easy for someone to figure out how to use the driver/flutter methods.

KazuCocoa replied:
Sure.

My intention was if both had the same purpose to check the same python client functionality, e2e just took longer time than the unit test and flakiness, so just it is redundant as checking.

It's fine to keep it. Then, could you leave somewhere when this e2e has an issue, who could we mention...? In `flutter_e2e_test` section is fine. (or just mention MummanaSubramanya?) Main maintainers in python client have no knowledge about the drier. We may need to mention you when this e2e driver fails if such failure is the driver side issue or client.

For example usage, I'd recommend you have such example in the driver project and run it every driver update to detect possible changes in the driver side that effects client implementation. Not main topic here but more general thing

MummanaSubramanya replied:
@KazuCocoa I completely agree with your concern about driver issue. As you suggested I can mention my name in the workflow file.

Agreed with your recommendation to have the e2e tests in a seperate standalone project. I can do it in a different PR if you want.

KazuCocoa replied:
thank you for the comment :)

MummanaSubramanya commented:
@KazuCocoa @mykola-mokhnach Thank you very much for your review. I have addressed all the comments and added both unit tests and end-to-end tests. I'm happy to report that all tests have passed.

mykola-mokhnach commented on appium/webdriver/extensions/flutter_integration/flutter_commands.py:
```diff
@@ -0,0 +1,194 @@
+    def wait_for_visible(
+        self,
+        locator: Union[WebElement, FlutterFinder],
+        time_out: Optional[float] = None,
```
we have timeout everywhere. Is there any specific reason to rename the var to time_out?

MummanaSubramanya replied:
It is only used in two places: wait_for_visible and wait_for_invisible. Could you clarify the issue with the time_out name? I believe it clearly indicates its purpose for the wait

mykola-mokhnach replied:
there is no issue. It's just called `timeout` (without underscore everyhere), so it would make sense to preserve this style

MummanaSubramanya replied:
Thanks! Updated :)

mykola-mokhnach reviewed (approved):
thanks @MummanaSubramanya

KazuCocoa commented:
It is not released yet. I'll release this tonight

KazuCocoa commented:
4.2.0 is available now
